vowpalwabbit/cats_pdf.cc

```diff
@@ -0,0 +1,189 @@
+  {
+    // add output example
+    reduction_output::report_progress(all, data, ec);
+    reduction_output::output_predictions(all.final_prediction_sink, ec.pred.prob_dist);
+    VW::finish_example(all, ec);
+  }
+
+  void reduction_output::output_predictions(
+    std::vector<std::unique_ptr<VW::io::writer>>& predict_file_descriptors, actions_pdf::pdf& prediction)
+  {
+    // output to the prediction to all files
+    const std::string str = to_string(prediction, true);
+    for (auto& f : predict_file_descriptors)
+      f->write(str.c_str(), str.size());
+  }
+
+  // "average loss" "since last" "example counter" "example weight"
+  // "current label" "current predict" "current features"
+  void reduction_output::report_progress(vw& all, cats_pdf&, example& ec)
+  {
+    const auto& cb_cont_costs = ec.l.cb_cont.costs;
+    all.sd->update(ec.test_only, does_example_have_label(ec), cb_cont_costs.empty() ? 0.f : cb_cont_costs[0].cost,
+        ec.weight, ec.num_features);
+    all.sd->weighted_labels += ec.weight;
+    print_update_cb_cont(all, ec);
+  }
+
+  inline bool reduction_output::does_example_have_label(example& ec)
+  {
+    return (!ec.l.cb_cont.costs.empty() && ec.l.cb_cont.costs[0].action != FLT_MAX);
+  }
+
+  void reduction_output::print_update_cb_cont(vw& all, example& ec)
+  {
+    if (all.sd->weighted_examples() >= all.sd->dump_interval && !all.logger.quiet && !all.bfgs)
+    {
+      all.sd->print_update(all.holdout_set_off, all.current_pass,
+          to_string(ec.l.cb_cont.costs[0]),  // Label
+          to_string(ec.pred.prob_dist),      // Prediction
+          ec.num_features, all.progress_add, all.progress_arg);
+    }
+  }
+
+  // END: functions to output progress
+  ////////////////////////////////////////////////////
+
+  // Setup reduction in stack
+  LEARNER::base_learner* setup(config::options_i& options, vw& all)
+  {
+    option_group_definition new_options("Continuous action tree with smoothing");
+    int num_actions = 0, pdf_num_actions = 0, cats_tree_actions = 0;
+    new_options.add(make_option("cats_pdf", num_actions).keep().help("Continuous action tree with smoothing (pdf)"))
+        .add(make_option("pmf_to_pdf", pdf_num_actions).keep().help("Convert pmf to pdf"))
+        .add(make_option("cats_tree", cats_tree_actions).keep().help("Continuous action tree"));
+
+    options.add_and_parse(new_options);
+
+  // If cats reduction was not invoked, don't add anything
+    // to the reduction stack;
+    if (!options.was_supplied("cats_pdf"))
+      return nullptr;
+
+    if (num_actions <= 0)
+      THROW(error_code::num_actions_gt_zero_s);
+
+    // cats stack = [cats_pdf -> cb_explore_pdf -> pmf_to_pdf -> get_pmf -> cats_tree ... rest specified by cats_tree]
+    if (!options.was_supplied("cb_explore_pdf"))
+      options.insert("cb_explore_pdf", "");
+    if (!options.add_or_check_options("pmf_to_pdf", num_actions))
+      THROW(error_code::options_disagree_s);
```

Comment:
I think we should consider not polluting the error_codes with errors that happen on the layer of command line arguments. Shouldn't they be more for runtime errors?

This might be a bigger conversation (regarding vw as a whole) but in my mind I see something like this:

CLI args / config file / json -> validate input step -> all ok -> do actual setup

Reply:
setup functions have several responsibilities today.  1) to check if a reduction should be activated 2) validate the arguments 3) specify reductions dependencies 4) instantiate the reduction using options specified by the user (or other setup functions) 5) create a learner that's next on the stack.

In the long term it makes sense to encapsulate these.  For now, this seems like a good place to do (2) and report errors.